elephant/gpfa.py

```diff
@@ -0,0 +1,407 @@
+        Data structure of test dataset.
+        Default is None.
+
+    Returns
+    -------
+
+    params_est : dict
+        Estimated model parameters, including `Corth`, obtained by
+        orthonormalizing the columns of C.
+    seqs_train : np.recarray
+        Training data structure that contains the new field `xorth`,
+        the orthonormalized neural trajectories.
+    seqs_test : np.recarray
+        Test data structure that contains orthonormalized neural
+        trajectories in `xorth`, obtained using `params_est`.
+        When no test dataset is given, None is returned.
+
+
+    Raises
+    ------
+    ValueError
+        If `fit_info['kernSDList'] != kern_sd`.
+
+    """
+    C = params_est['C']
+    X = np.hstack(seqs_train['xsm'])
+    Xorth, Corth, _ = gpfa_util.orthogonalize(X, C)
+    seqs_train = gpfa_util.segment_by_trial(seqs_train, Xorth, 'xorth')
+
+    params_est['Corth'] = Corth
+
+    if seqs_test is not None:
+        print("Extracting neural trajectories for test data...\n")
+        seqs_test = gpfa_core.exact_inference_with_ll(seqs_test,
+                                                      params_est)
+        X = np.hstack(seqs_test['xsm'])
+        Xorth, Corth, _ = gpfa_util.orthogonalize(X, C)
+        seqs_test = gpfa_util.segment_by_trial(seqs_test, Xorth, 'xorth')
+
+    return params_est, seqs_train, seqs_test
+
+
+def gpfa(data, bin_size=20*pq.ms, x_dim=3, em_max_iters=500):
+    """
+    Prepares data and calls functions for extracting neural trajectories.
+
+    Parameters
+    ----------
+
+    data : list
+          list of spike trains in different trials
+                                        0-axis --> Trials
+                                        1-axis --> Neurons
+                                        2-axis --> Spike times
+    bin_size : quantities.Quantity, optional
+        Width of each time bin.
+        Default is 20 ms.
+    x_dim : int, optional
+        State dimensionality.
+        Default is 3.
+    em_max_iters : int, optional
+        Number of EM iterations to run (default: 500).
+
+    Returns
+    -------
+
+    parameter_estimates: dict
+        Estimated model parameters.
+        When the GPFA method is used, following parameters are contained
+            covType: {'rbf', 'tri', 'logexp'}
+                type of GP covariance
+                Currently, only 'rbf' is supported.
+            gamma: ndarray of shape (1, #latent_vars)
+                related to GP timescales by 'bin_width / sqrt(gamma)'
+            eps: ndarray of shape (1, #latent_vars)
+                GP noise variances
+            d: ndarray of shape (#units, 1)
+                observation mean
+            C: ndarray of shape (#units, #latent_vars)
+                mapping between the neuronal data space and the latent variable
+                space
+            R: ndarray of shape (#units, #latent_vars)
+                observation noise covariance
+
+    seqs_train: numpy.recarray
```

Comment:
I am not sure I understand the advantage of using `recarray`  instead of a classic list of `dict` here? Would
```python
seqs_train['y'][0] # recarrays, trial 0
```
be preferrable over
```python
seqs_train[0]['y'] # list of dicts, trial 0
```
from a user perspective?

Reply:
I fully agree not using `np.recarray`. The authors should provide a clear reason why they stick to np.recarray over python-friendly structures. Apart from confusing the users (and probably developers as well), it has several disadvantages during the test setup and future maintenance.

Reply:
I also though to define `seqs_train` as a dict rather than a recarray, but I found that there are many part where `seqs_train['y']` is subjected to vector operation, which is not quickly done with a dict.
We could still define it as a dict of arrays (of the same length, i.e., # of trials) representing `y`, `x`, and so on, but it would require a total rewrite of the whole code.

By the way, for a recarray `seqs_train`, `seqs_train['y'][0]` is identical to `seqs_train[0]['y']`.